### PR TITLE
Further improve RRDP metrics.

### LINKED ISSUE #477: Feature: RRDP serial as a metric
As a repository operator, to be sure that my repository is getting updates, I like to monitor that the RRDP serial is increasing.  For this I would like to have the RRDP serial as a Prometheus metric.

```yaml
name: SerialReset
expr: rrdp_serial{address=~".*[host]/.*"} < max_over_time(rrdp_serial{address=~".*[host]/.*"}[30m])
labels:
severity: P4
annotations:
description: {{ $labels.group }} RRDP serial at {{ $labels.address }} reset from {{ with printf "max_over_time(rrdp_serial{address='%s'}[30m])" .Labels.address | query }}{{ . | first | value }}{{end}} to {{ $value }} (in the last 30m) according to {{ $labels.instance }}.
summary: {{ $labels.group }} RRDP serial at {{ $labels.address }} reset to {{ $value }}.
```

For bonus points, track accepted/rejected rrdp updates as well, such as OctoRPKI does:
```yaml
name: OctoRPKIRRDPErrors
expr: rrdp_errors{address="https://[host]/rrdp/notification.xml"} > 0
labels:
severity: P1
annotations:
description: OctoRPKI found {{ humanize $value }} RRDP errors for {{ $labels.address} from {{ $labels.instance }}.
summary: OctoRPKI found {{ humanize $value }} RRDP errors for {{ $labels.address}} on {{ $labels.hostname }}
```

```yaml
name: OctoRPKIRRDPDeltaIncreasing
expr: (increase(rrdp_serial{address=~".*[host]/.*"}[1h30m]) == 0)
labels:
severity: P1
annotations:
description: OctoRPKI did see {{ $value }} new RRDP serials at {{ $labels.address }} during the last 90  minutes from {{ $labels.instance }}.
summary: No new RRDP serial at {{ $labels.address }} for 90m.
```

### PR DESCRIPTION
**Pull request by partim**
This PR further enhances the RRDP repository metrics and resolves #477.

This changes the meaning of ‘status’ to be the status code of the last HTTP request made with separate status information available for the notification file and the series of payload requests. In addition, if a snapshot was requested, the reason for that is given.

@ties, is it enough to have the new combined status in the Prometheus or do you rather have more Prometheus metrics and if so which?

### CODE REVIEW

**Comment by ties** [comment]
Hi,

This is a nice addition for my use case.

I tried to build this branch (clone, `cargo build`, `./target/debug/routinator -c ~/.rpki-cache/config -vv server`) but I don't see some of the metrics that I would expect based on the changes and the description.

However I do see that the `routinator_rrdp_status` metric sometimes turns to 1 when there is no failed request in the logs. What does this value mean?
```
# HELP routinator_rrdp_status status code for getting notification file
# TYPE routinator_rrdp_status gauge
routinator_rrdp_status{uri="https://rrdp.ripe.net/notification.xml"} -1
...

# HELP routinator_rrdp_serial serial number of last RRDP update
# TYPE routinator_rrdp_serial gauge
routinator_rrdp_serial{uri="https://rrdp.ripe.net/notification.xml"} 669
```
The rrdp serial is perfect for my use case (detecting liveliness of publication points) :).

**Comment by partim** [comment]
The missing metrics aren’t in the Prometheus metrics but only in the /status and /api/v1/status endpoints. I probably could add at least notify-status and payload-status there as well, if that is useful.

*  *  *
A status code of -1 is supposed to mean that fetching the response failed, so probably an IO error. There should indeed be log output further explaining the issue. There might be a mistake in how the value is derived from notify-status and  payload-status. I’ll have a look.

**Comment by partim** [comment]
Ah! You get the -1 payload-status code when there was no payload request because the data was up-to-date. That seems suboptimal. I’ll figure something out.

**Comment by partim** [comment]
This should be supported now. The payload status is now 0 if there was no payload query necessary.

**Comment by ties** [comment]
I did not know about `/api/v1/status` :). The prometheus endpoint looks very good now and this definitely covers my use case.

File: src/http.rs [hunk 1]
```diff
@@ -838,9 +881,15 @@ fn api_status(origins: &OriginsHistory) -> Response<Body> {
...
                     target.member_raw("delta",
-                        if metrics.delta { "true" }
+                        if metrics.snapshot_reason.is_none() { "true" }
                         else { "false" }
                     );
+                    if let Some(reason) = metrics.snapshot_reason {
+                        target.member_str("snapshot_reason", reason.code())
+                    }
+                    else {
+                        target.member_raw("snapshot_reason", "none");
```
Comment (ties) [inline src/http.rs:891]: I don't think `none` is a valid json value

Comment (partim) [reply src/http.rs]: Argl. Fixed.

**Comment by partim** [comment]
`/api/v1/status` was introduced in 0.8.3 to supply status information to the UI.

**Review by density215 — APPROVED** [review]

**Pull request merged**